crates/solvers/src/infra/dex/balancer/mod.rs

```diff
@@ -0,0 +1,151 @@
+//! Bindings to the Balancer Smart Order Router (SOR) API.
+
+use {
+    crate::{
+        boundary,
+        domain::{auction, dex, eth, order},
+    },
+    ethereum_types::U256,
+    std::sync::atomic::{self, AtomicU64},
+};
+
+mod dto;
+
+pub struct Sor {
+    client: reqwest::Client,
+    endpoint: reqwest::Url,
+    vault: boundary::balancer::Vault,
+    settlement: eth::ContractAddress,
+}
+
+pub struct Config {
+    /// The URL for the Balancer SOR API.
+    pub endpoint: reqwest::Url,
+
+    /// The address of the Balancer Vault contract.
+    pub vault: Option<eth::ContractAddress>,
+
+    /// The address of the Settlement contract.
+    pub settlement: Option<eth::ContractAddress>,
+}
+
+impl Sor {
+    pub async fn swap(
+        &self,
+        order: &dex::Order,
+        slippage: dex::Slippage,
+        gas_price: auction::GasPrice,
+    ) -> Result<dex::Swap, Error> {
+        let query = dto::Query::from_domain(order, gas_price);
+        let quote = {
+            // Set up a tracing span to make debugging of API requests easier.
+            // Historically, debugging API requests to external DEXs was a bit
+            // of a headache.
+            static ID: AtomicU64 = AtomicU64::new(0);
+            let id = ID.fetch_add(1, atomic::Ordering::Relaxed);
+            let span = tracing::trace_span!("swap", id = %id);
+            let _enter = span.enter();
```

Comment:
This only works for synchronous code:
> Warning: In asynchronous code that uses async/await syntax, Span::enter may produce incorrect traces if the returned drop guard is held across an await point. See [the method documentation](https://docs.rs/tracing/latest/tracing/span/struct.Span.html#in-asynchronous-code) for details.

For async functions this would do what you expect:
```rust
async {}.instrument(span);
```

This would also be compatible with my next suggestion: It might be nice to have the tracing span for the entire `swap()` function. That way we could also associate any new logs that are related to (de)serializing request/response (should we ever add them).

Reply:
Nice catch!

Reply:
> It might be nice to have the tracing span for the entire `swap()` function.

I intentionally only set up a span around the `POST /quote` API request as a way to link the request and response objects.

My general strategy for these single order solvers would be to have a span with `{order = "...uid..."}` for each attempted to solve order, making it easier to debug exactly what happened to a specific order. Also worth noting that, in theory for a partially fillable order, you might have multiple concurrent Balancer SOR requests for a single order (in order to find a partial fill amount).

Reply:
Fixed the instrumentation issue in 0bcffecd411b53910a72df8d7b67955c567747ff. Created issue to remember to do a general "instrument `solvers` crate" pass in #1239